python/ql/src/experimental/Security/CWE-352/DjangoCSRF.ql

```diff
@@ -0,0 +1,36 @@
+/**
+ * @name Disabled Django CSRF protection
+ * @description No csrf Middleware/ Decorator to protect Django application from CSRF
+ * @kind problem
+ * @problem.severity error
+ * @precision high
+ * @id py/django-csrf
+ * @tags security
+ *       external/cwe/cwe-352
+ */
+
+import python
+import semmle.python.strings
+
+from GlobalVariable g
+where
+  /*
+   * Regex matching MIDDLEWARE & MIDDLEWARE_CLASSES(backwards compatibility)
+   * Checks if there is a GlobalVariable(List/Tuple) named MIDDLEWARE or MIDDLEWARE_CLASES
+   * and contains the middleware string
+   */
+
+  g.getId().regexpMatch("MIDDLEWARE(_CLASSES)?") and
+  not exists(List l |
+    l.getParentNode().getAChildNode().toString().regexpMatch(g.getId()) and
```

Comment:
I'm not happy about this way of identifying that we're assigning `g` to `l`.

Have a go at playing with this snippet, and rewriting your query so you don't need to duplicate the logic for both List and Tuple.

```codeql
import python

from GlobalVariable g, Assign assign, Expr elt
where
    g.getId() = "MIDDLEWARE" and
    assign.getATarget() = g.getAStore() and
    (
        elt = assign.getValue().(List).getAnElt()
        or
        elt = assign.getValue().(Tuple).getAnElt()
    )
select g, assign, elt

```

Reply:
This makes a lot of sense :D  modified the query